analyses/cell-type-ewings/scripts/run-infercnv.R

```diff
@@ -0,0 +1,160 @@
+# define output metadata file 
+cnv_metadata_file <- file.path(opt$output_dir, glue::glue("{library_id}_cnv-metadata.tsv"))
+
+# define output infercnv obj
+scratch_obj_file <- file.path(scratch_dir, "run.final.infercnv_obj")
+output_obj_file <- file.path(opt$output_dir, glue::glue("{library_id}_cnv-obj.rds"))
+
+# png file to save 
+scratch_png <- file.path(scratch_dir, "infercnv.png")
+output_png <- file.path(opt$output_dir, glue::glue("{library_id}_infercnv.png"))
+
+# Define normal cells ----------------------------------------------------------
+
+if(!is.null(opt$reference_cell_file)){
+  
+  # make sure normal cells file exists 
+  stopifnot("reference file does not exist" = file.exists(opt$reference_cell_file))
+  normal_cells <- readr::read_tsv(opt$reference_cell_file) |> 
+    dplyr::filter(reference_cell_class == "Normal") |> 
+    dplyr::pull(barcodes)
+  
+  # check that all normal cells are in colnames of sce 
+  stopifnot("All barcodes in the normal_cells file are not found in the sce object" = 
+              all(normal_cells %in% colnames(sce)))
+  
+  # define reference 
+  ref_name <- "reference"
+  
+} else {
+  # otherwise set normal cells to default 
+  normal_cells <- ""
+  # define reference 
+  ref_name <- NULL
+}
+
+# pull out normal cells and create annotations
+annotation_df <- data.frame(barcodes = colData(sce)$barcodes) |> 
+  dplyr::mutate(annotations = dplyr::if_else(
+    barcodes %in% normal_cells,
+    "reference",
+    "unknown"
+  )) 
+
+readr::write_tsv(annotation_df, opt$annotations_file, col_names = FALSE)
+
+# Run inferCNV -----------------------------------------------------------------
+
+# create the infercnv object
+infercnv_obj <- infercnv::CreateInfercnvObject(raw_counts_matrix=counts(sce),
+                                               annotations_file=opt$annotations_file,
+                                               delim="\t",
+                                               gene_order_file=opt$gene_order_file,
+                                               ref_group_names=ref_name)
+
+# run infercnv and report by cell rather than group
+infercnv_obj <- infercnv::run(infercnv_obj,
+                              cutoff=0.1,  # use 1 for smart-seq, 0.1 for 10x-genomics
+                              out_dir=scratch_dir, # save all intermediate files to scratch dir
+                              denoise=T,
+                              HMM=T ,
+                              save_rds = FALSE, # don't save the intermediate rds files
+                              num_threads = opt$threads
+)
+
+# Save final results -----------------------------------------------------------
+
+# create table with barcodes and CNVs for each chromosome
+infercnv::add_to_seurat(seurat_obj = NULL,
+                                      infercnv_output_path = scratch_dir)
+
+# define path to output cnv metadata file saved in scratch dir
+scratch_metadata_file <- file.path(scratch_dir, "map_metadata_from_infercnv.txt")
```

Comment:
define this path at the top?

I also wonder if we want to keep these named by library, as with all the other files.

Reply:
We can't name this file. This is just defining the path to the file that's saved by InferCNV and then when we copy it over we add the library ID.

Reply:
Yeah, sorry, I was confused!